### PR TITLE
fix(httpclient): apply transport wrappers in layer order at Build

### PR DESCRIPTION
**Pull request by gaborage**
## Problem

`httpclient.Builder` holds exactly **one** transport field. Every `With*` option that needs a `RoundTripper` wraps whatever is in that field **at call time**, so the last call wins and becomes outermost. Today only `WithJOSE` does this, and the resulting footgun was merely *documented*: calling `WithTransport` after `WithJOSE` silently discarded the JOSE layer.

Four queued payment-integration transports — #765 (field-level encryption), #766 (OAuth 1.0a signing), #767 (mTLS), #768 (x-pay-token HMAC) — each add another wrapper to that same field. Written independently, they produce a client whose behavior depends on the order the caller happened to call the builder, and two failure modes are silent and security-relevant:

- A client-certificate transport installed **after** a body-transform wrapper is discarded, so the client connects with **no client certificate**.
- A wrapper whose `Inner` is nil falls back to `net/http.DefaultTransport` — again dialing with no client certificate, no error, no log line.

## Change

Wrappers now register into **ordered layers** and are applied at `Build()` time, innermost-first, regardless of call order:

- `layerSigner` (inner) → `layerBodyTransform` (outer). Signers sit beneath body transforms so a signature covers the bytes **actually placed on the wire** — the ordering all four queued transports depend on.
- `transportChain.byLayer` is indexed by layer, so the array index *is* the apply order and there is no separate order list to keep in sync. `layerCount` is an iota sentinel, so a layer declared above it is applied with no second edit elsewhere.
- Ordering within a layer is registration order, first registered innermost.

**Zero new exported identifiers**, so `apidiff` is strictly neutral — no ADR, no `wiki/migrations.md` hop. The chain hangs off a **pointer** field (`chain *transportChain`) specifically so `Builder` stays a comparable type; a bare slice field would make the exported `Builder` non-comparable, which `apidiff` reports as INCOMPATIBLE.

### Also included

`Build()` now emits a WARN for the one composition that silently downgrades TLS: a registered wrapper with **no** `WithTransport` base replaces the `Transport` of a client passed to `WithHTTPClient`, so requests dial via `net/http.DefaultTransport` — losing client certificates, pinned `RootCAs`, `MinVersion` and proxy settings. This is pre-existing behavior (byte-for-byte identical trigger set before and after this change), but this PR adds documentation that markets composition safety, so shipping the guardrail alongside it is the honest move. Failing *closed* there needs an API change (`Build()` has no error return) and stays with the mTLS work in **#767**.

## Behavior notes

- **Intentional behavior change, no API change.** Callers who relied on `WithTransport`-after-`WithJOSE` wiping the JOSE layer now get both. That reliance is implausible and the old behavior was a documented footgun, but it *is* observable. Hence `fix(httpclient):` — not `feat:`, not `fix(...)!:`.
- `WithTransport(nil)` after a registered wrapper no longer *clears* it; it yields wrapper-over-nil-base. This is the only input where the new code returns non-nil where the old returned nil.
- `Build()` mints a fresh `*JOSETransport` per call where the old code reused one instance across repeated `Build()`s. `JOSETransport` is stateless, so this is strictly better isolation.
- `httpclient/jose_transport.go` is **byte-identical** — verified in CI-equivalent checks.

## Verification

- `make check` green (fmt + lint + `-race` across all packages + alloc guards + govulncheck 0 vulns).
- 12 new tests/subtests. **Both halves of the fix are mutation-checked**: reversing layer order fails `TestBuilderTransportChainLayerOrdering` while `…OrderIndependence` stays green (clean isolation, proving the two tests pin different mechanisms); reverting the `Build()` resolution collapses both.
- The five pre-existing `TestBuilder` transport subtests pass **unmodified** — the change to `client_test.go` is machine-verified append-only, so none of them was quietly edited to pass.
- `golangci-lint` (incl. gosec) 0 issues on the changed package; no secrets, no raw-SQL escape hatches in the diff.

## Follow-ups this unblocks

#766 and #768 register at `layerSigner`; #765 registers at `layerBodyTransform` alongside JOSE; #767 supplies the base via `WithTransport` (or a future `WithClientCertificate`). Each should add a composition test asserting its wrapper lands at the right depth, and none should reintroduce wrap-at-call-time.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Transport wrappers now compose deterministically, independent of the order builder options are applied.
  * JOSE processing integrates correctly with custom transport and signing layers.
  * Added warning when wrapper layers replace a provided client transport in ways that may drop custom TLS/proxy settings.

* **Documentation**
  * Added guidance on transport composition order and correct configuration (including “WRONG/RIGHT” examples).

### CODE REVIEW

File: httpclient/client.go [hunk 1]
```diff
@@ -264,8 +335,14 @@ func (b *Builder) Build() Client {
...
-		httpClient.Transport = b.transport
+	if rt := b.resolveTransport(); rt != nil {
+		if b.discardsClientTransport() && b.logger != nil {
+			b.logger.Warn().Msg("httpclient: transport wrapper registered without WithTransport — " +
+				"the *http.Client passed to WithHTTPClient has its own Transport replaced and requests " +
+				"dial via net/http.DefaultTransport, losing client certificates, pinned roots and proxy " +
+				"settings; pass the base RoundTripper to WithTransport instead")
+		}
+		httpClient.Transport = rt
 	}
```
[the remark this answers is not in the document]
Comment (gaborage) [reply httpclient/client.go:346]: Addressed — but by **removing** a guard rather than adding one.

Verified first: the lines this depends on are untouched by the PR. `logger: b.logger` in `Build()` and the unconditional `c.logger.Info()` / `c.logger.Debug()` call sites are all pre-existing (base commit `9a46cfa`, lines 273 / 1099 / 1114 / 1145 / 1159), and `git diff 9a46cfa -- httpclient/client.go | grep -c 'logger:               b.logger'` returns `0`. So the nil-logger panic predates this PR and isn't introduced by it.

What this PR *did* introduce was an inconsistency: a `b.logger != nil` guard on the new warn path, while every other logger use in the file dereferences unconditionally. That guard implied a nil-tolerant contract the client doesn't actually honor — a caller passing `nil` still panics on the first request regardless. Removed it, so the new code matches the file's existing assumption instead of contradicting it.

Deliberately **not** fixing the underlying contract here:
- Installing a no-op logger would silently swallow every log line for a misconfigured caller — against this repo's "no silent failures" rule (CLAUDE.md, Robustness).
- Rejecting `nil` in `NewBuilder` changes the runtime behavior of a shipped exported constructor.

Both deserve their own change with their own tests, rather than riding along in an apidiff-neutral transport-composition PR. Flagged to the maintainer as a follow-up candidate.

File: httpclient/client_test.go [hunk 2]
```diff
@@ -1823,3 +1823,142 @@ func TestClientDoTransportErrorSpan(t *testing.T) {
...
+	t.Run("warned_case_really_replaces_the_transport", func(t *testing.T) {
+		built := NewBuilder(log).WithHTTPClient(withClient()).WithJOSE(JOSEConfig{}).Build()
+		clientImpl, ok := built.(*client)
+		require.True(t, ok)
+		assert.NotSame(t, base, clientImpl.httpClient.Transport, "caller's transport is replaced, not wrapped")
+		joseTransport, ok := clientImpl.httpClient.Transport.(*JOSETransport)
+		require.True(t, ok)
+		assert.Nil(t, joseTransport.Inner, "no base: RoundTrip falls back to net/http.DefaultTransport")
+	})
+}
```
[the remark this answers is not in the document]
Comment (gaborage) [reply httpclient/client_test.go:1974]: Applied. Good catch — the test pinned the warning's *side effect* but not the warning itself.

Added a `warnSpy` that embeds `logger.Logger` / `logger.LogEvent` and overrides only `Warn()` and `Msg()` — the two methods `Build()` actually calls — plus two subtests:

- `hazard_emits_the_warning` — asserts the emitted message on the `WithHTTPClient(...).WithJOSE(...)` path.
- `explicit_base_emits_no_warning` — asserts silence when `WithTransport` supplies a base, so the spy can't pass by always reporting a warning.

Interface embedding keeps the double to ~10 lines: this repo has no log-capture seam (no logger constructor accepts an `io.Writer`, and no existing test anywhere asserts an emission), so a hand-written full mock would have meant ~20 stub methods across both interfaces.

The nil-logger-safety half of the suggestion is moot — I removed that guard rather than testing it; see the reply on `httpclient/client.go`.
